design-decisions/2018-10-12-Condition-Split.md

```diff
@@ -0,0 +1,96 @@
+Splitting configurations upon new constraints on symbolic variables
+===================================================================
+
+Background
+----------
+
+When applying an axiom on a symbolic configuration new constraints may be
+introduced on symbolic variables, coming from side conditions, or directly
+from the unification process.
+
+### Example (from side conditions)
+
+Consider rewriting term `3 / A:Int` using
+
+`rule X:Int / Y:Int => X /Int Y requires Y =/=Int 0`
+which would be written as
+`Y =/= 0 = true /\ X:Int / Y:Int => X /Int Y`
+
+Then `3 / A:Int /\ (Y =/=Int 0 = true /\ X:Int / Y:Int)` will simplify to
+```
+3 / A:Int /\ (X:Int = 3 /\ Y:Int = A:Int)
+          /\ (A:Int =/=Int 0) = true
+```
+which is satisfiable, but not `T`.
+
+### Example (from unification)
+
+Consider rewriting term `if A:Int <Int 0 then -1 else 1` using
+
+`rule if true then E1:Exp else E2:Exp => E1`
+
+Then `if A:Int <Int 0 then -1 else 1 /\ if true then E1:Exp else E2:Exp` will
+simplify to
+```
+if A:Int <Int 0 then -1 else 1 /\ (E1:Exp = -1 /\ E2:Exp = 1)
+                               /\ A:Int <Int 0 = true
+```
+which is satisfiable, but not `T`.
+
+Problem
+-------
+
+Simply continuing the rewriting process when a new constraint is introduced
+(on some symbolic variable) is not sound when the constraint does not simplify
+to top.  The reason is that, going further, the coverage is reduced to the
+inputs which satisfy the constraint.
```

Comment:
I don't think this is a good description of the problem, this whole thing only comes up as an issue if we are not handling evaluation in a sound way! If the rule above was the only applicable rule, then it *is true* that `((3 / A:int) /\ \next(E2:Exp)) <->((A:Int =/= 0)  = true /\ (E2 = 3t /Int A:Int)`.

The ideas in this proposal are ok as a means of trying to keep the configuration in an easily-understood form, but if soundness is ever at risk that's a warning that we are not following the principle of expressing everything in logically sound ways.

Reply:
Even if the above rule isn't the only applicable one, we can still claim a sound one-path step as follows.
* Grab the unification result `(X:Int = 3 /\ Y:Int = A:Int) /\ (A:Int =/=Int 0) = true`.
* Existentially quantify free variables in the rule, i.e., `X` and `Y`.
* Simplify. We get `(A:Int =/=Int 0) = true`.
* Construct the term after one-step rewriting `3 /Int A /\ (A:Int =/=Int 0) = true`.

And, it is indeed that
```
|- (exists X Y . (X / Y /\ Y =/=Int 0)) /\ (3 / A => 3 /Int A /\ (A:Int =/=Int 0) = true)
```

Reply:
I think the problem here is the fact that in the above claim, we have `(exists X Y . (X / Y /\ Y =/=Int 0))` as part of the lhs of the one-step reachability claim. Intuitively, it means the set of all patterns where the rule
`rule X:Int / Y:Int => X /Int Y requires Y =/=Int 0`
is applicable. 

I guess what @traiansf wanted to point out is that, if these "constraints of applicability", denoted `NC`, are not equivalent to `\top`, then continuing the process will end up with the following claim:
```
|- NC1 /\ NC2 /\ ... /\ NCk /\ (CT /\ CC) => CT'
```
where the accumulative constraints, `NC1 /\ NC2 /\ ... /\ NCk`, may only cover a very small portion of the input space.

Reply:
I'd say as long as `NC1 /\ NC2 /\ ... /\ NCk` is not unsatisfiable, then we're in a good shape, if one-path reachability is considered. The search/branch should be cut off if `NC1 /\ NC2 /\ ... /\ NCk` is unsatisfiable.